src/db.c

```diff
@@ -1319,11 +1319,12 @@ int expireIfNeeded(redisDb *db, robj *key) {
  * pre-allocated keysbuf or by allocating a new array on the heap.
  */
 static int *prepareGetKeysResult(getKeysResult *result, int numkeys) {
-    long unsigned int size = sizeof(int) * numkeys;
-    if (size > sizeof(result->keysbuf))
-        result->keys = zmalloc(size);
-    else
+    if (numkeys > result->size) {
+        result->keys = zmalloc(numkeys * sizeof(int));
+        result->size = numkeys;
+    } else {
         result->keys = result->keysbuf;
+    }
```

Comment:
i think it's better to set the `size` member in the else too?
instead or in addition to the one in `GETKEYS_RESULT_INIT`.

also, maybe we want to support this function to be called repeatedly, in which case it may need to release the old pointer.